controllers/front/OrderConfirmationController.php

```diff
@@ -58,17 +61,21 @@ public function init()
         $this->secure_key = Tools::getValue('key', false);
         $order = new Order((int) ($this->id_order));
 
-        if (!$this->id_order || !$this->id_module || !$this->secure_key || empty($this->secure_key)) {
+        if (!$this->id_order || (!$this->isFreeOrder && !$this->id_module) || !$this->secure_key || empty($this->secure_key)) {
             Tools::redirect($redirectLink . (Tools::isSubmit('slowvalidation') ? '&slowvalidation' : ''));
         }
         $this->reference = $order->reference;
         if (!Validate::isLoadedObject($order) || $order->id_customer != $this->context->customer->id || $this->secure_key != $order->secure_key) {
             Tools::redirect($redirectLink);
         }
-        $module = Module::getInstanceById((int) ($this->id_module));
-        if ($order->module != $module->name) {
-            Tools::redirect($redirectLink);
+
+        if (!$this->isFreeOrder) {
```

Comment:
```suggestion
        if (!$order->currentOrder) {
```

Reply:
No, I think we should keep check for `free` order